### PR TITLE
python3Packages.riscof: init 1.25.2

### PR DESCRIPTION
**Pull request by GenericNerdyUsername**
###### Description of changes

Add riscof (and dependencies)

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Relevant: #213222

### CODE REVIEW

**Comment by Et7f3** [comment]
Did you applied suggestion on the other PR ?

File: pkgs/applications/virtualization/sail-riscv/default.nix
Comment (NickCao) [inline pkgs/applications/virtualization/sail-riscv/default.nix:4]: Remove these extraneous empty lines between arguments.

File: pkgs/applications/virtualization/sail-riscv/default.nix
Comment (NickCao) [inline pkgs/applications/virtualization/sail-riscv/default.nix:21]: Multiline strings are more readable.
```suggestion
  prePatch = ''
    rm -r prover_snapshots
  '';
```

File: pkgs/applications/virtualization/sail-riscv/default.nix
Comment (NickCao) [inline pkgs/applications/virtualization/sail-riscv/default.nix:30]: Keep them one per line. This would make future modifications easier and cleaner.
And is sail used as a tool during build? That way it should be in nativeBuildInputs.
```suggestion
  makeFlags = [
    "SAIL=sail"
    "ARCH=${arch}"
    "SAIL_DIR=${ocamlPackages.sail}/share/sail"
  ];
```

File: pkgs/applications/virtualization/sail-riscv/default.nix
Comment (NickCao) [inline pkgs/applications/virtualization/sail-riscv/default.nix:39]: Keep an empty line between the boilerplate codes and your own.
```suggestion
  installPhase = ''
    runHook preInstall

    mkdir -p $out/bin
    cp c_emulator/riscv_sim_${arch} $out/bin
    mkdir $out/share/
    cp -r generated_definitions/{coq,hol4,isabelle} $out/share/

    runHook postInstall
  '';
```

File: pkgs/development/ocaml-modules/lem/default.nix
Comment (NickCao) [inline pkgs/development/ocaml-modules/lem/default.nix:6]:
```suggestion
```

File: pkgs/development/ocaml-modules/lem/default.nix
Comment (NickCao) [inline pkgs/development/ocaml-modules/lem/default.nix:24]:
```suggestion
  installFlags = [ "INSTALL_DIR=$(out)" ];
```

File: pkgs/development/ocaml-modules/lem/default.nix
Comment (NickCao) [inline pkgs/development/ocaml-modules/lem/default.nix:27]: Also use multiline string here, even if it's just a single line.

File: pkgs/development/ocaml-modules/lem/default.nix
Comment (NickCao) [inline pkgs/development/ocaml-modules/lem/default.nix:29]: Src should be right below pname and version.

File: pkgs/development/ocaml-modules/linksem/default.nix
Comment (NickCao) [inline pkgs/development/ocaml-modules/linksem/default.nix:4]:
```suggestion
```

File: pkgs/development/ocaml-modules/sail/default.nix
Comment (NickCao) [inline pkgs/development/ocaml-modules/sail/default.nix:33]: Also keep them one per line.

File: pkgs/development/ocaml-modules/sail/default.nix
Comment (NickCao) [inline pkgs/development/ocaml-modules/sail/default.nix:59]: As in previous reviews, these manual overrides might not be needed.

File: pkgs/development/python-modules/dataproperty/default.nix
Comment (NickCao) [inline pkgs/development/python-modules/dataproperty/default.nix:21]: Writing deselect three times won't hurt. Keeping the code clean is more important.

File: pkgs/development/python-modules/dataproperty/default.nix
Comment (NickCao) [inline pkgs/development/python-modules/dataproperty/default.nix:18]: pytestCheckHook should be in nativeCheckInputs.

File: pkgs/development/python-modules/dataproperty/default.nix
Comment (NickCao) [inline pkgs/development/python-modules/dataproperty/default.nix:27]: Same here.

File: pkgs/development/python-modules/mbstrdecoder/default.nix
Comment (NickCao) [inline pkgs/development/python-modules/mbstrdecoder/default.nix:4]:
```suggestion
```

File: pkgs/development/python-modules/mbstrdecoder/default.nix
Comment (NickCao) [inline pkgs/development/python-modules/mbstrdecoder/default.nix:19]: Same.

**Review by NickCao — COMMENTED** [review]
Quite a lot formatting and code style issues. You can use nixpkgs-format, statix and other tools you help you locate/fix these issues.

File: pkgs/development/ocaml-modules/sail/default.nix
Comment (GenericNerdyUsername) [reply pkgs/development/ocaml-modules/sail/default.nix:59]: The build fails without them, as the default `buildDunePackage` only builds the `[pname]` package, but sail includes multiple packages in one source tree

**Comment by GenericNerdyUsername** [comment]
All the problems mentioned in the review should be fixed now

File: pkgs/development/python-modules/pytablewriter/default.nix
Comment (NickCao) [inline pkgs/development/python-modules/pytablewriter/default.nix:35]: setuptools shall not be propagated in most cases as it is a build-time only dependency. And it might not be required in the first place, as the default for `buildPythonPackage` is to treat the package as setuptools based, adding necessary dependencies.

File: pkgs/development/python-modules/riscv-isac/default.nix
Comment (NickCao) [inline pkgs/development/python-modules/riscv-isac/default.nix:4]:
```suggestion
```

File: pkgs/development/python-modules/riscv-isac/pyelftools-version.patch
Comment (NickCao) [inline pkgs/development/python-modules/riscv-isac/pyelftools-version.patch:9]: use `pythonRelaxDepsHook`

File: pkgs/development/python-modules/typepy/default.nix
Comment (NickCao) [inline pkgs/development/python-modules/typepy/default.nix:4]:
```suggestion
```

File: pkgs/development/python-modules/typepy/default.nix
Comment (NickCao) [inline pkgs/development/python-modules/typepy/default.nix:9]:
```suggestion
```

File: pkgs/development/python-modules/riscof/make_writeable.patch
Comment (GenericNerdyUsername) [reply pkgs/development/python-modules/riscof/make_writeable.patch:9]: It copies a template from the store and attempts to modify it later

File: pkgs/development/python-modules/riscof/gitpython-version.patch
Comment (GenericNerdyUsername) [reply pkgs/development/python-modules/riscof/gitpython-version.patch:10]: Am I missing something or does it not work with test dependencies?

Comment (NickCao) [reply pkgs/development/python-modules/riscof/gitpython-version.patch:10]: What do you mean?

File: pkgs/development/python-modules/riscof/make_writeable.patch
Comment (NickCao) [reply pkgs/development/python-modules/riscof/make_writeable.patch:9]: Sounds reasonable, please add a comment to clarify this.

File: pkgs/development/python-modules/riscof/gitpython-version.patch
Comment (GenericNerdyUsername) [reply pkgs/development/python-modules/riscof/gitpython-version.patch:10]: the build succeeds, but the tests fail due to a missing dependency. Looking at the docs, it seems the hook modifies the end result, not requirements.txt, which means we still need a way to modify test dependencies, and i cant think of a better way than this

Comment (NickCao) [reply pkgs/development/python-modules/riscof/gitpython-version.patch:10]: Then just use `substituteInPlace`, patches are harder to update later.

**Comment by GenericNerdyUsername** [comment]
Assuming fixing the merge conflict didn't break anything, I think this is ready to merge (unless there's something in the review i missed?)

**Review by NickCao — APPROVED** [review]
looks good to my eye, just need someone who knows the ocaml ecosystem better than me to take a second look.

**Comment by NickCao** [comment]
@vbgl sorry for pinging you without asking for permission, but you seem to be working with ocaml things recently and I hope to get a review from you.

**Comment by vbgl** [comment]
No problem. I’ll try to give a look at the OCaml parts of this PR.

File: pkgs/development/ocaml-modules/lem/default.nix
Comment (vbgl) [inline pkgs/development/ocaml-modules/lem/default.nix:14]:
```suggestion
  pname = "ocaml${ocaml.version}-lem";
```

File: pkgs/development/ocaml-modules/lem/default.nix
Comment (vbgl) [inline pkgs/development/ocaml-modules/lem/default.nix:25]: This is not used (you have to check & `throw` yourself).

File: pkgs/development/ocaml-modules/lem/default.nix
Comment (vbgl) [inline pkgs/development/ocaml-modules/lem/default.nix:34]:
```suggestion
  createFindlibDestdir = true;
```

File: pkgs/development/ocaml-modules/lem/default.nix
Comment (vbgl) [inline pkgs/development/ocaml-modules/lem/default.nix:44]: The `platforms` attribute seems missing.

File: pkgs/development/ocaml-modules/lem/default.nix
Comment (vbgl) [inline pkgs/development/ocaml-modules/lem/default.nix:28]: Why isn’t `num` also propagated?

File: pkgs/development/ocaml-modules/lem/default.nix
Comment (vbgl) [inline pkgs/development/ocaml-modules/lem/default.nix:27]: Is `ncurses` really needed?

**Review by vbgl — COMMENTED** [review]
Most of the comments/suggestion also apply to the `linksem` library.

File: pkgs/development/ocaml-modules/lem/default.nix
Comment (GenericNerdyUsername) [reply pkgs/development/ocaml-modules/lem/default.nix:27]: no, i dont know why i data that (or why i put `num` in there either)

File: pkgs/development/ocaml-modules/lem/default.nix
Comment (vbgl) [reply pkgs/development/ocaml-modules/lem/default.nix:25]: Here is an example: https://github.com/NixOS/nixpkgs/blob/9313ae3ba50422fb5b916103dcca9d77898264fa/pkgs/development/ocaml-modules/uuidm/default.nix#L3-L4

`minimalOCamlVersion` is handled by `buildDunePackage` only.

File: pkgs/development/ocaml-modules/lem/default.nix
Comment (GenericNerdyUsername) [reply pkgs/development/ocaml-modules/lem/default.nix:44]: I'm unable to test anything other than `x86_64-linux` without emulation, is it ok if that's the only platform I put there?

**Comment by GenericNerdyUsername** [comment]
@vbgl can you check if i missed any of your recommendations?

File: pkgs/development/ocaml-modules/linksem/default.nix
Comment (vbgl) [inline pkgs/development/ocaml-modules/linksem/default.nix:24]:
```suggestion
```

File: pkgs/development/ocaml-modules/lem/default.nix
Comment (vbgl) [inline pkgs/development/ocaml-modules/lem/default.nix:28]:
```suggestion
```

**Comment by GenericNerdyUsername** [comment]
Done

**Comment by NickCao** [comment]
Result of `nixpkgs-review pr 213832` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>20 packages built:</summary>
  <ul>
    <li>python310Packages.dataproperty</li>
    <li>python310Packages.mbstrdecoder</li>
    <li>python310Packages.pytablewriter</li>
    <li>python310Packages.riscof</li>
    <li>python310Packages.riscv-config</li>
    <li>python310Packages.riscv-isac</li>
    <li>python310Packages.tabledata</li>
    <li>python310Packages.tcolorpy</li>
    <li>python310Packages.typepy</li>
    <li>python311Packages.dataproperty</li>
    <li>python311Packages.mbstrdecoder</li>
    <li>python311Packages.pytablewriter</li>
    <li>python311Packages.riscof</li>
    <li>python311Packages.riscv-config</li>
    <li>python311Packages.riscv-isac</li>
    <li>python311Packages.tabledata</li>
    <li>python311Packages.tcolorpy</li>
    <li>python311Packages.typepy</li>
    <li>sail-riscv-rv32</li>
    <li>sail-riscv-rv64</li>
  </ul>
</details>

**Comment by NickCao** [comment]
@GenericNerdyUsername thanks for your contribution and patience in addressing review comments.
@vbgl also thanks for your review.

**Pull request merged**
